coalib/misc/Caching.py

```diff
@@ -0,0 +1,122 @@
+        >>> cache.write()
+
+        >>> time.sleep(1)
+
+        Let's create a new instance to simulate a separate run:
+        >>> cache = FileCache(log_printer, "coala_test", flush_cache=False)
+
+        >>> old_data = copy.deepcopy(cache.last_cache())
+
+        We can mark a file as changed by doing:
+        >>> cache.add_to_changed_files({"a.c"})
+
+        >>> cache.write()
+        >>> cache = FileCache(log_printer, "coala_test", flush_cache=False)
+        >>> new_data = cache.last_cache()
+
+        Since we marked 'a.c' as a changed file:
+        >>> old_data["a.c"] == new_data["a.c"]
+        True
+
+        Since 'b.c' was untouched after the second run, its time was updated
+        to the latest value:
+        >>> old_data["b.c"] < new_data["b.c"]
+        True
+
+        :param log_printer: A LogPrinter object to use for logging.
+        :param project_dir: The root directory of the project to be used
+                            as a key identifier.
+        :param flush_cache: Flush the cache and rebuild it.
+        """
+        self.log_printer = log_printer
+        self.project_dir = project_dir.encode("utf-8")
+        self.md5sum = hashlib.md5(self.project_dir).hexdigest()
+        self.current_time = calendar.timegm(time.gmtime())
+        if not time_consistent(log_printer, self.md5sum):
+            log_printer.warn("It seems like you went back in time - your "
+                             "system time is behind the last recorded run "
+                             "time on this project. The cache will "
+                             "be flushed and rebuilt.")
+            flush_cache = True
+        if not flush_cache:
+            self.data = pickle_load(log_printer, self.md5sum, {})
+        else:
+            self.data = {}
+            delete_cache_files(log_printer, [self.md5sum])
+            log_printer.info("The file cache was successfully flushed.")
+        self.changed_files = set()
+
+    def write(self):
+        """
+        Update the last run time on the project for each file
+        to the current time.
+        """
+        for file_name in self.data:
+            if (self.data[file_name] == -1 or
+                    file_name not in self.changed_files):
+                self.data[file_name] = self.current_time
+        pickle_dump(self.log_printer, self.md5sum, self.data)
+        update_time_db(self.log_printer, self.md5sum, self.current_time)
+        self.changed_files = set()
+
+    def add_to_changed_files(self, changed_files):
+        """
+        Keep track of changed files in ``changed_files`` for future use in
+        ``write``.
+
+        :param changed_files: A set of files that had changed since the last
+                              run time.
+        """
+        self.changed_files.update(changed_files)
+
+    def last_cache(self):
+        """
+        Get the last run time on the project in epoch format
+        (number of seconds since Jan 1, 1970).
```

Comment:
Just wanted to check here whether the epoch is right for Windows too ?
Normally the windows epoch is not the 1970 date, it's something else (I'm sure Microsoft Excel uses a different epoch, 1990 I think ?).